### ISSUE TITLE
Node configuration shouldn't be order-specific

### ISSUE THREAD

**Issue by jarohen** [xtdb/xtdb#1644]
_Originally posted by @cmgrote in https://github.com/xtdb/xtdb/issues/1588#issuecomment-942347356_

On advice from @refset I managed to workaround this in my case by using an EDN configuration rather than a JSON configuration; however, there seems to then be a subsequent quirk where if you've defined the `:xtdb/document-store` portion before other portions (like the transaction log or a Lucene configuration), then even though the `xtdb/document-store` portion is parsed and configured those subsequent options do not seem to be parsed or configured.

In other words: only when the `:xtdb/document-store` portion is at the very end of the EDN configuration does all of the configuration get parsed and processed as part of standing up a node.  (And this ordering-specificity appears to _only_ occur when using the JDBC configuration for the document store, other configurations like kv-stores don't appear to behave the same way.)  So I'm assuming the two issues, since they're both JDBC-config-specific, may be related (?)

**Comment by jarohen** [comment]
https://github.com/xtdb/xtdb/issues/1588#issuecomment-943161639

That's weird, EDN shouldn't be order-specific! My initial thought would be that, depending on the size of the map, Clojure sometimes opts for an ArrayMap like implementation, sometimes a HashMap. The system startup turns a partial ordering into a total ordering, so if it's an ArrayMap it may choose a different total ordering, but I would have thought the partial ordering should deal with this. :thinking:

@cmgrote - have you got a reproducible example you could share?

**Comment by cmgrote** [comment]
Sure, this config:

```clojure
{
  :xtdb/index-store {
    :kv-store {
      :xtdb/module xtdb.rocksdb/->kv-store
      :db-dir "data/rdb-index"
    }
  }
  :xtdb/document-store {
    :xtdb/module xtdb.jdbc/->document-store
    :connection-pool {
      :dialect {:xtdb/module xtdb.jdbc.psql/->dialect}
      :db-spec {:jdbcUrl "jdbc:postgresql://localhost/db?user=login&password=password"}
    }
  }
  :xtdb/tx-log {
    :kv-store {
      :xtdb/module xtdb.rocksdb/->kv-store
      :db-dir "data/rdb-tx"
    }
  }
  :xtdb.lucene/lucene-store {:db-dir "data/lucene"}
}
```

Seems to miss off the configuration of both the transaction log and Lucene piece (no files created, and looking at the debug-level output of XTDB starting up the node Lucene doesn't appear to be configured at all).

However, if we just change around the order to this, then everything appears to be configured (and directories created) for i.e. Lucene and tx log:

```clojure
{
  :xtdb/index-store {
    :kv-store {
      :xtdb/module xtdb.rocksdb/->kv-store
      :db-dir "data/rdb-index"
    }
  }
  :xtdb/tx-log {
    :kv-store {
      :xtdb/module xtdb.rocksdb/->kv-store
      :db-dir "data/rdb-tx"
    }
  }
  :xtdb.lucene/lucene-store {:db-dir "data/lucene"}
  :xtdb/document-store {
    :xtdb/module xtdb.jdbc/->document-store
    :connection-pool {
      :dialect {:xtdb/module xtdb.jdbc.psql/->dialect}
      :db-spec {:jdbcUrl "jdbc:postgresql://localhost/db?user=login&password=password"}
    }
  }
}
```

My hunch would be that it isn't so much that ordering is important, but that whatever hiccup is causing problems with the `jdbcUrl` parsing in other areas may be causing the startup to "drop-out" silently during / just after the `jdbcUrl` portion -- so if this shows up before others (due to the array ordering you mention) then the other pieces of the config are inadvertently skipped, while if it is at the very end then whatever is silently crashing out doesn't end up skipping the earlier-ordered config options (?)

**Comment by jarohen** [comment]
Hey @cmgrote, thanks for the config :pray: Unfortunately I can't repro it here - here's what I'm doing:

* Put the top config in your above message into `/tmp/config.edn`, change Postgres creds to point to a locally running Postgres Docker container
* Against 1.19.0, calling `IXtdb.startNode(new File("/tmp/config.edn"))`
* Node starts up, creates 3 directories under `./data`

Also, if I keep the Postgres URL as-is (i.e. so that it's unable to connect), I get the exception thrown as expected (rather than the node silently starting up without it).

Is this what you're doing/seeing?

Cheers,

James
